src/ring_theory/witt_vector/domain.lean

```diff
@@ -0,0 +1,209 @@
+/-!
+From this point on, we assume `R` is of characteristic `p`.
+-/
+
+section char_p
+
+variable [char_p R p]
+
+lemma iterate_verschiebung_mul_frobenius (x : 𝕎 R) (i j : ℕ) :
+  (frobenius^[i] : 𝕎 R → 𝕎 R) (verschiebung^[j] x) =
+    (verschiebung^[j] ((frobenius^[i] x))) :=
+iterate_comm_apply (λ _, (verschiebung_frobenius_comm _).symm) _ _ _
+
+lemma iterate_verschiebung_mul (x y : 𝕎 R) (i j : ℕ) :
+  (verschiebung^[i] x) * (verschiebung^[j] y) =
+    (verschiebung^[i + j] ((frobenius^[j] x) * (frobenius^[i] y))) :=
+begin
+  calc
+  _ = (verschiebung^[i] (x * (frobenius^[i] ((verschiebung^[j] y))))) : _
+... = (verschiebung^[i] (x * (verschiebung^[j] ((frobenius^[i] y))))) : _
+... = (verschiebung^[i] ((verschiebung^[j] ((frobenius^[i] y)) * x))) : _
+... = (verschiebung^[i] ((verschiebung^[j] ((frobenius^[i] y) * (frobenius^[j] x))))) : _
+... = (verschiebung^[i + j] ((frobenius^[i] y) * (frobenius^[j] x))) : _
+... = _ : _,
+  { apply iterate_verschiebung_mul_left },
+  { rw iterate_verschiebung_mul_frobenius },
+  { rw mul_comm },
+  { rw iterate_verschiebung_mul_left },
+  { rw iterate_add_apply },
+  { rw mul_comm }
+end
+
+lemma iterate_frobenius_coeff (x : 𝕎 R) (i k : ℕ) :
+  ((frobenius^[i] x)).coeff k = (x.coeff k)^(p^i) :=
+begin
+  induction i with i ih,
+  { simp },
+  { rw [iterate_succ_apply', coeff_frobenius_char_p, ih],
+    ring_exp }
+end
+
+/-- This is a slightly specialized form of [Hazewinkel, *Witt Vectors*][Haze09] 6.2 equation 5. -/
+lemma iterate_verschiebung_mul_coeff (x y : 𝕎 R) (i j : ℕ) :
+  ((verschiebung^[i] x) * (verschiebung^[j] y)).coeff (i + j) =
+    (x.coeff 0)^(p ^ j) * (y.coeff 0)^(p ^ i) :=
+begin
+  calc
+  _ = (verschiebung^[i + j] ((frobenius^[j] x) * (frobenius^[i] y))).coeff (i + j) : _
+... = ((frobenius^[j] x) * (frobenius^[i] y)).coeff 0 : _
+... = (frobenius^[j] x).coeff 0 * ((frobenius^[i] y)).coeff 0 : _
+... = _ : _,
+  { rw iterate_verschiebung_mul },
+  { convert iterate_verschiebung_coeff _ _ _ using 2,
+    rw zero_add },
+  { apply coeff_mul_zero },
+  { simp only [iterate_frobenius_coeff] }
+end
+
+/-!
+## Witt vectors over a domain
+
+If `R` is an integral domain, then so is `𝕎 R`.
+This argument is adapted from <https://tinyurl.com/2p8cwrn7>.
+-/
+
+variable  [is_domain R]
+
+lemma eq_zero_or_eq_zero_of_mul_eq_zero (x y : 𝕎 R) : x * y = 0 → x = 0 ∨ y = 0 :=
+begin
+  contrapose!,
+  rintros ⟨ha, hb⟩,
+  rcases verschiebung_nonzero ha with ⟨na, wa, hwa0, hwaeq⟩,
+  rcases verschiebung_nonzero hb with ⟨nb, wb, hwb0, hwbeq⟩,
+  have : (x * y).coeff (na + nb) = (wa.coeff 0) ^ (p ^ nb) * (wb.coeff 0) ^ (p ^ na),
+  { rw [← iterate_verschiebung_mul_coeff, hwaeq, hwbeq], },
+  have : (x * y).coeff (na + nb) ≠ 0,
+  { rw this,
+    apply mul_ne_zero; apply pow_ne_zero; assumption },
+  contrapose! this,
+  simp [this]
+end
+
+instance : is_domain (𝕎 R) :=
+{ eq_zero_or_eq_zero_of_mul_eq_zero := eq_zero_or_eq_zero_of_mul_eq_zero,
+  exists_pair_ne := witt_vector.nontrivial.exists_pair_ne }
```

Comment:
More generally true as:
```suggestion
instance [no_zero_divisors R] : no_zero_divisors (𝕎 R) :=
⟨λ x y, begin
  contrapose!,
  rintros ⟨ha, hb⟩,
  rcases verschiebung_nonzero ha with ⟨na, wa, hwa0, hwaeq⟩,
  rcases verschiebung_nonzero hb with ⟨nb, wb, hwb0, hwbeq⟩,
  have : (x * y).coeff (na + nb) = (wa.coeff 0) ^ (p ^ nb) * (wb.coeff 0) ^ (p ^ na),
  { rw [← iterate_verschiebung_mul_coeff, hwaeq, hwbeq], },
  have : (x * y).coeff (na + nb) ≠ 0,
  { rw this,
    apply mul_ne_zero; apply pow_ne_zero; assumption },
  contrapose! this,
  simp [this]
end⟩

instance [is_domain R] : is_domain (𝕎 R) :=
{ ..witt_vector.no_zero_divisors,
  ..witt_vector.nontrivial }
```

Reply:
(although I guess `char_p` probably implies `nontrivial` anyway)

Reply:
Actually, you can golf the first one to:
```
instance [no_zero_divisors R] : no_zero_divisors (𝕎 R) :=
⟨λ x y, begin
  contrapose!,
  rintros ⟨ha, hb⟩,
  rcases verschiebung_nonzero ha with ⟨na, wa, hwa0, rfl⟩,
  rcases verschiebung_nonzero hb with ⟨nb, wb, hwb0, rfl⟩,
  refine ne_of_apply_ne (λ x, x.coeff (na + nb)) _,
  rw [iterate_verschiebung_mul_coeff, zero_coeff],
  refine mul_ne_zero (pow_ne_zero _ hwa0) (pow_ne_zero _ hwb0),
end⟩
```
which you could always add some type annotations to to make more readable.